src/functionalities/Functionality.js

```diff
@@ -12,19 +12,34 @@ class Functionality {
       slots: Joi.array().items(Joi.alternatives().try(
         ...SlotFactory.SUPPORTED_CLASSES_SCHEMAS,
       )).required(),
+      controllerName: Joi.string().allow(''), // this is used to support the old devices: https://github.com/ospin-web-dev/hambda/issues/913
     })
   }
 
+  _addSlot(slotData) {
+    const { id: functionalityId } = this
+    const newSlot = SlotFactory.new({ ...slotData, functionalityId })
+
+    this.slots.push(newSlot)
+
+    return newSlot
+  }
+
+  _addSlots(slotsData) {
+    /* NOTE: slots should never be added or removed outside of initialization
+     * a.k.a. if I could make this method more private I would */
+    slotsData.map(slotData => this._addSlot(slotData))
+  }
+
   constructor({
     id,
     name,
-    slots,
+    slots: slotsData,
   }) {
     this.id = id
     this.name = name
-    this.slots = Array.isArray(slots)
-      ? slots.map(SlotFactory.new)
-      : []
+    this.slots = []
```

Comment:
I recall a best practice when initializing objects is to always provide everything the object will need in its lifecycle in the constructor (e.g., avoid doing things like adding a new `this.property` in some helper method because that makes the instance a lot less transparent). For this reason are empty slots added here, then immediately populated in the add slots call.